examples/wmt/train.py

```diff
@@ -220,17 +229,35 @@ def loss_fn(params):
                                                       label_smoothing)
     mean_loss = loss / weight_sum
     return mean_loss, logits
-
+  optimizer = train_state.optimizer
   step = optimizer.state.step
   lr = learning_rate_fn(step)
-  grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
-  (_, logits), grad = grad_fn(optimizer.target)
-  grad = jax.lax.pmean(grad, "batch")
-  new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+  if train_state.dynamic_scale:
+    grad_fn = train_state.dynamic_scale.value_and_grad(
+        loss_fn, has_aux=True, axis_name="batch")
+    dynamic_scale, is_fin, aux, grads = grad_fn(optimizer.target)
+    train_state = train_state.replace(dynamic_scale=dynamic_scale)
+    # dynamic loss takes care of averaging gradients across replicas
```

Comment:
oh-- i guess this question it out of scope for this PR, but why is it like this? is it because you need to do the pmean first to detect nans anyways, so it must happen within DynamicScale?

Reply:
Each device must either skip or not skip so pmean makes sure that either every or no device has a NaN